src/Ardalis.Specification/Evaluators/SearchMemoryEvaluator.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System.Diagnostics;
+
+namespace Ardalis.Specification;
+
+public class SearchMemoryEvaluator : IInMemoryEvaluator
+{
+    private SearchMemoryEvaluator() { }
+    public static SearchMemoryEvaluator Instance { get; } = new SearchMemoryEvaluator();
+
+    public IEnumerable<T> Evaluate<T>(IEnumerable<T> query, ISpecification<T> specification)
+    {
+        if (specification.SearchCriterias is List<SearchExpressionInfo<T>> { Count: > 0 } list)
+        {
+            // The search expressions are already sorted by SearchGroup.
+            return new SpecLikeIterator<T>(query, list);
+        }
+
+        return query;
+    }
+
+    private sealed class SpecLikeIterator<TSource> : Iterator<TSource>
+    {
+        private readonly IEnumerable<TSource> _source;
+        private readonly List<SearchExpressionInfo<TSource>> _searchExpressions;
+
+        private IEnumerator<TSource>? _enumerator;
+
+        public SpecLikeIterator(IEnumerable<TSource> source, List<SearchExpressionInfo<TSource>> searchExpressions)
```

Comment:
Not switching to primary constructor syntax yet?

Reply:
I've been hesitant because of the older TFMs. We need to check whether we need any polyfills or not (I assume not).